inc/Addon/Cloudflare/API/Client.php

```diff
@@ -0,0 +1,203 @@
+<?php
+declare(strict_types=1);
+
+namespace WP_Rocket\Addon\Cloudflare\API;
+
+use Exception;
+use WP_Rocket\Addon\Cloudflare\Auth\AuthInterface;
+use WP_Rocket\Addon\Cloudflare\Auth\CredentialsException;
+
+class Client {
+	const CLOUDFLARE_API = 'https://api.cloudflare.com/client/v4/';
+
+	/**
+	 * Auth object
+	 *
+	 * @var AuthInterface
+	 */
+	private $auth;
+
+	/**
+	 * An array of arguments for wp_remote_request()
+	 *
+	 * @var array
+	 */
+	protected $args = [];
+
+	/**
+	 * Constructor.
+	 *
+	 * @param AuthInterface $auth Auth implementation.
+	 */
+	public function __construct( AuthInterface $auth ) {
+		$this->auth = $auth;
+		$this->args = [
+			'sslverify' => true,
+			'body'      => [],
+			'headers'   => [],
+		];
+	}
+
+	/**
+	 * API call method for sending requests using GET.
+	 *
+	 * @param string $path Path of the endpoint.
+	 * @param array  $data Data to be sent along with the request.
+	 *
+	 * @return object
+	 */
+	public function get( $path, array $data = [] ) {
+		return $this->request( 'get', $path, $data );
+	}
+
+	/**
+	 * API call method for sending requests using POST.
+	 *
+	 * @param string $path Path of the endpoint.
+	 * @param array  $data Data to be sent along with the request.
+	 *
+	 * @return object
+	 */
+	public function post( $path, array $data = [] ) {
+		return $this->request( 'post', $path, $data );
+	}
+
+	/**
+	 * API call method for sending requests using DELETE.
+	 *
+	 * @param string $path Path of the endpoint.
+	 * @param array  $data Data to be sent along with the request.
+	 *
+	 * @return object
+	 */
+	public function delete( $path, array $data = [] ) {
+		return $this->request( 'delete', $path, $data );
+	}
+
+	/**
+	 * API call method for sending requests using PATCH.
+	 *
+	 * @param string $path Path of the endpoint.
+	 * @param array  $data Data to be sent along with the request.
+	 *
+	 * @return object
+	 */
+	public function patch( $path, array $data = [] ) {
+		return $this->request( 'patch', $path, $data );
+	}
+
+	/**
+	 * API call method for sending requests
+	 *
+	 * @param string $method Type of method that should be used.
+	 * @param string $path   Path of the endpoint.
+	 * @param array  $data   Data to be sent along with the request.
+	 *
+	 * @return object
+	 *
+	 * @throws Exception When empty content.
+	 * @throws AuthenticationException When email or api key are not set.
+	 * @throws UnauthorizedException When Cloudflare's API returns a 401 or 403.
+	 * @throws CredentialsException When credentials are not valid.
+	 */
+	protected function request( $method = 'get', $path, array $data = [] ) {
+		try {
+			if (
+				'/ips' !== $path
+				&&
+				! $this->auth->is_valid_credentials()
+			) {
+				throw new AuthenticationException( 'Authentication information must be provided.' );
```

Comment:
@Tabrisrp I am wondering if the usage of an exception base logic is based on something (maybe exceptions in a library you use? ) as it looks like in the plugin we try to reduce as much as possible the usage of exceptions.

Reply:
I kept the usage of exception that we had in the previous implementation, but no since we're not using library for this, there is no constraints on that. We could replace them by using WP_Error objects instead if we want